sql/delegate: include composite types in SHOW TYPES

spilchen opened the pull request:
SHOW TYPES previously only included the enumerated types. In 23.1 composite types were added, but we forgot to update SHOW TYPES. This corrects that by including composite types in the catalog lookup.

Unlike other SHOW statements, there is no way to filter out schema. So, this will return composite types for all schemas. We do a lookup to ResolveSchema, knowing it won't find a schema, in order to get the current database name.

I query pg_catalog.pg_type to get the types. But there are a bunch of internal types in their that need to be omitted:
- anything from any of the known internal schemas are skipped: information_schema, pg_catalog, crdb_internal, and pg_extension.
- each time a table is created, a type for the table descriptor is added. I query typrelid to filter out the table descriptor. It must be 0, for enums, or refer back to its own oid so that we know its a standalone type.

Fixes: #119734
Release note (bug fix): SHOW TYPES includes user defined composite types. It omitted those types ever since composite types were added in 23.1.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124730)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
nice work! the fix looks good to me, but i just left a minor comment on how to do the database lookup

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-:-Nyzestz-JhJvB8HU_Hy:bw4iw5z)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/delegate/show_types.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-NyzeCenEp0cKvCorBHo:-NyzeCenEp0cKvCorBHp:bhwaw6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/12cbacc1adc65a064561f399ffa65929985a2dc9/pkg/sql/delegate/show_types.go#L27)):*
> ```Go
> 	// call ResolveSchema with an empty schema to specifically retrieve current
> 	// database/schema name.
> 	_, name, err := d.catalog.ResolveSchema(d.ctx, flags, &cat.SchemaName{})
> ```

nit: we can access the current database in a slightly easier way that avoids performing a lookup. instead, we can look at the in-memory session data, like this: https://github.com/cockroachdb/cockroach/blob/48233b40f4191b88738b06daf4bcf86dced77fad/pkg/sql/delegate/show_grants.go#L235
___
*[`pkg/sql/logictest/testdata/logic_test/composite_types` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-NyzerK8AlU3D2GC-Fi3:-NyzerK8AlU3D2GC-Fi4:b-j6rd9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/12cbacc1adc65a064561f399ffa65929985a2dc9/pkg/sql/logictest/testdata/logic_test/composite_types#L274)):*
> ```
> public  et1   root
> sc1     ct2   root
> sc1     ct3   root
> ```

nice test!


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/delegate/show_types.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-NyzeCenEp0cKvCorBHo:-Nz-5X8m1-Z74DmoIlo2:bndjwm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/12cbacc1adc65a064561f399ffa65929985a2dc9/pkg/sql/delegate/show_types.go#L27)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we can access the current database in a slightly easier way that avoids performing a lookup. instead, we can look at the in-memory session data, like this: https://github.com/cockroachdb/cockroach/blob/48233b40f4191b88738b06daf4bcf86dced77fad/pkg/sql/delegate/show_grants.go#L235
</blockquote></details>

Much simpler.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/delegate/show_types.go` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-Nz-gDlf9B3vkBWNKVCA:-Nz-gDlgCuhJT9EdoJ6q:b-k410tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdb0fd54e710ce3bc2590a22918dda3009983fa8/pkg/sql/delegate/show_types.go#L42)):*
> ```Go
>       nsp.nspname NOT IN ('information_schema', 'pg_catalog', 'crdb_internal', 'pg_extension')
> ORDER BY (nsp.nspname, types.typname)
> `, d.evalCtx.SessionData().Database)
> ```

i just remembered another tricky edge case that matters here. if the name of an identifier has upper-case letters or if the name has certain characters such as spaces, `"`, or `-`, then the name needs to be double-quoted for it to be parsed correctly. there are two ways this can be achieved:
- `tree.Name(d.evalCtx.SessionData().Database).String()`
- `lexbase.EscapeSQLIdent(d.evalCtx.SessionData().Database)`

The first option just ends up calling in to the second option. I think within the `delegate` package, the first option might be more conventional to use.

a test could be added for this in the logic tests with something like:
```
CREATE DATABASE "MyDatabase";
USE "MyDatabase";
CREATE TYPE ct5 AS (a INT, b TEXT); 
CREATE TYPE ct6 AS ENUM ('a', 'b', 'c');

SHOW TYPES
```

sorry for not remembering this earlier!


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/delegate: include composite types in SHOW TYPES

SHOW TYPES previously only included the enumerated types. In 23.1
composite types were added, but we forgot to update SHOW TYPES. This
corrects that by including composite types in the catalog lookup.

I query pg_catalog.pg_type to get the types. But there are a bunch of
internal types in their that need to be omitted:
- anything from any of the known internal schemas are skipped:
  information_schema, pg_catalog, crdb_internal, and pg_extension.
- each time a table is created, a type for the table descriptor is
  added. I query typrelid to filter out the table descriptor. It must be
  0, for enums, or refer back to its own oid so that we know its a
  standalone type.

Fixes: #119734
Release note (bug fix): SHOW TYPES includes user defined composite
types. It omitted those types ever since composite types were added in
23.1.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/delegate/show_types.go` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-Nz-gDlf9B3vkBWNKVCA:-Nz3Ae1P1YxF-5Mw5lFA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdb0fd54e710ce3bc2590a22918dda3009983fa8/pkg/sql/delegate/show_types.go#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i just remembered another tricky edge case that matters here. if the name of an identifier has upper-case letters or if the name has certain characters such as spaces, `"`, or `-`, then the name needs to be double-quoted for it to be parsed correctly. there are two ways this can be achieved:
- `tree.Name(d.evalCtx.SessionData().Database).String()`
- `lexbase.EscapeSQLIdent(d.evalCtx.SessionData().Database)`

The first option just ends up calling in to the second option. I think within the `delegate` package, the first option might be more conventional to use.

a test could be added for this in the logic tests with something like:
```
CREATE DATABASE "MyDatabase";
USE "MyDatabase";
CREATE TYPE ct5 AS (a INT, b TEXT); 
CREATE TYPE ct6 AS ENUM ('a', 'b', 'c');

SHOW TYPES
```

sorry for not remembering this earlier!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! when you're ready feel free to merge using `bors`

this change is a bug fix, so it's also something we should backport to older release branches, so the fix goes out in patch releases for existing versions. we should backport to 23.1, 23.2, and 24.1, since those are the current supported releases.

this page describes 3 different options for how to do a backport: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/900005932/Backporting+a+change+to+a+release+branch. (the last option is required if there are any merge conflicts.) also, we can discuss this more during our 1-1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124730#-:-Nz3pl8l6j1pFOqxpPVY:bh3wky0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
bors r=rafiss
